version: "3.9"
services:
  db:
    image: postgres:15.1
    volumes:
      - app-db-data:/var/lib/postgresql/data/
    environment:
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_DB: ${POSTGRES_DB}
    ports:
      - "5432:5432"
    networks:
      - app-network
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -U postgres -d schedule" ]
      interval: 5s
      timeout: 5s
      retries: 5

  backend:
    build:
      context: .
    depends_on:
      db:
        condition: service_healthy
    volumes:
      - .:/app
    ports:
      - 8000:${BACKEND_PORT}
    environment:
      DEBUG: ${DEBUG}

      BACKEND_TITLE: ${BACKEND_TITLE}
      BACKEND_DESCRIPTION: ${BACKEND_DESCRIPTION}
      BACKEND_PREFIX: ${BACKEND_PREFIX}

      BACKEND_HOST: ${BACKEND_HOST}
      BACKEND_PORT: ${BACKEND_PORT}
      BACKEND_RELOAD: ${BACKEND_RELOAD}

      POSTGRES_SERVER: ${POSTGRES_SERVER}
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_DB: ${POSTGRES_DB}

      ENV: development
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.fastapi.rule=Host(`fastapi.localhost`)"
    networks:
      - app-network

  traefik:
    image: traefik:v2.2
    ports:
      - "8008:80"
    networks:
      - app-network
    volumes:
      - "./traefik.dev.toml:/etc/traefik/traefik.toml"
      - "/var/run/docker.sock:/var/run/docker.sock:ro"

volumes:
  app-db-data:

networks:
  app-network: